_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,340 @@
+
+With at-least-once, there is no way for the producer to know the nature of the failure, so it simply assumes that the message was not written and retries.
+There are use cases in which lost or duplicated messages are acceptable, but in others this would cause serious business problems (e.g. a duplicate debit from a bank account).
+
+To be fair, there is no such thing as exactly-once delivery in distributed systems ([two generals' problem](https://en.wikipedia.org/wiki/Two_Generals%27_Problem)), but we can fake it by implementing idempotent log append.
+Another way to achieve EOS is to apply a deduplication logic at the consumer side, but this is a stateful logic that needs to be replicated in every consumer.
+
+> Idempotence is the property of certain operations in mathematics and computer science whereby they can be applied multiple times without changing the result beyond the initial application.
+
+Since Kafka 3.0, the producer enables the stronger delivery guarantees by default (`acks=all`, `enable.idempotence=true`).
+In addition to durability, this provides partition level message ordering and duplicates protection.
+
+When the idempotent producer is enabled, the broker registers the producer id (PID) and the epoch (producer session).
+A sequence number is assigned to each record when the batch is first added to a produce request and never changed, even if the batch is resent.
+The broker keeps track of the sequence number per producer and partition, periodically storing this information in a `.snapshot` file.
+Whenever a new batch arrives, the broker checks if the last received number is equal to the first batch number plus one, then the batch is acknowledged, otherwise it is rejected.
+
+Unfortunately, the idempotent producer does not guarantee duplicate protection across restarts and when you need to write to multiple partitions as a single unit of work.
+This is the typical scenario with many stream processing applications that run read-process-write cycles.
+In this case, you can make these cycles atomic using the transaction support of the Kafka Producer or Kafka Streams APIs.
+
+The Kafka Producer API is low level, so you need to ensure the right configuration and to commit offsets within the scope of the transaction.
+A transaction should be aborted in case failure during processing, which means none of the messages from the transaction will be readable by transactional consumers.
+Instead, with the Kafka Streams API you can simply set `processing.guarantee=exactly_once_v2` to enable EOS on your existing topology.
+
+### Considerations
+
+Each producer must configure its own static and unique [`transactional.id`](https://kafka.apache.org/documentation/#producerconfigs_transactional.id) to enable the zombie fencing logic, which avoids message duplicates.
+The `transactional.id` is used to uniquely identify the same producer instance across process restarts, and it is associated with the producer session.
```

Comment:
```suggestion
The `transactional.id` is used to uniquely identify the same logical producer across process restarts. In contrast to the idempotent producer, a transactional producer instance will be allocated _the same_ PID (but incremented producer epoch) as any previous instance with the same `transactional.id`. Together the PID and producer epoch identify a logical _producer session_.
```

Reply:
Thanks! This really makes the producer session concept clear.